### ISSUE TITLE
Wagtail Search as a separated package

### ISSUE THREAD

**Issue by Alexx-G** [wagtail/wagtail#3716]
Is it possible to have wagtail search as a separated/reusable package?

I mean, wagtail search has quite good implementation and seems to be reusable for a lot of other projects (with some adjustments indeed).
2-3 years ago I needed elasticsearch in one of projects I was working on and Wagtail's implementation was the best match. It was quite easy to get only needed modules from wagtail search and reuse them in that project. So, in my opinion it's possible to maintain wagtailsearch as a separated package/project so it might be reused for other projects.

What do you think?

**Comment by BertrandBordage** [comment]
I totally agree with you. And same goes for other Wagtail components, like the admin interface, StreamFields or RichTextFields. Ideally, we should be able to use all that without having to use a CMS and create the corresponding tables.

I think I can speak for the team by saying we definitely want to make Wagtail more modular and not necessarily focused on a tree structure of pages. But I don't think we will split Wagtail into multiple PyPI packages, at least in the next  2-3 years. It's way easier to create releases of a single monolithic package. It seems less elegant, we won't have to issue a dozen releases every two months, with a compatibility grid to explain what is compatible with what, and warnings for potential incompatibilities, etc.

To make it short, one day we will be able to use only wagtail search, but we will have to install the whole PyPI package of Wagtail for that. I don't think that's a problem, we install Django completely yet we rarely use more than 70 % of its components.

In my opinion, Wagtail should be a high level framework for making sites, just a level higher than Django. And for some points like the admin interface, Wagtail proposes a replacement for a Django component that's dying. By the way, I think Django admin will have to be removed from Django at some point. And Wagtail's admin could become the official replacement ;)

**Comment by Alexx-G** [comment]
Sorry for late response.
I understand your rationale and you're doing great job on improving Wagtail. And it makes sense to install wagtail if I'm going to replace django's admin with it and use the search package. 
However, if I need only search and I don't use admin (I have projects like this), it doesn't seem right to install a CMS (great indeed) only to get good integration with elasticsearch.

As a suggestion, you might isolate independent packages in separated repositories and link them into main repo using git submodules. In this way you don't have to deploy each package in PyPi, on the other hand, I can still install only that package using pip.

**Comment by sasha0** [comment]
> But I don't think we will split Wagtail into multiple PyPI packages, at least in the next 2-3 years.

As "next 2-3 years" has passed, raising the topic again 😉 .

I looking to rework search in the [Django-Oscar](https://github.com/django-oscar/django-oscar) and wondering if Wagtail search extraction would get into the further development plans.

**Comment by thibaudcolas** [comment]
This would be pretty cool. I’m going to convert this to Discussions as this seems to be the kind of big idea that’d require lots of work and thought just to validate the concept
